WordPress/gutenberg
# REST API: add WP_REST_Posts_Controller route to fetch page counts

## What?
Alternative to https://github.com/WordPress/gutenberg/pull/66294

Extends `WP_REST_Posts_Controller` with a new route to fetch page counts


This PR only affects the `page` post type. 

 In the Core sync, all post types that inherit from `WP_REST_Posts_Controller` will get the `/count` route.



## Why?

The need for a performant post counts check for the block editor. See: https://github.com/WordPress/gutenberg/pull/65223#issuecomment-2381323815

Contrast this with `$query->found_posts` which returns the total count for a specific query only and would degrade performance given `n` posts.

## How?
Via `wp/v2/pages/count`, the editor can fetch the total number of post counts. 

This is done using [wp_count_posts](https://developer.wordpress.org/reference/functions/wp_count_posts/).



## Testing Instructions

Run tests:

`npm run test:unit:php:base -- --filter Gutenberg_Test_REST_Posts_Controller`

Examples to run in the browser console (in the editor):

```js

await wp.apiFetch( { path: '/wp/v2/pages/count' } )

/*
{
    "publish": 1,
    "future": 0,
    "draft": 2,
    "pending": 0,
    "private": 0,
    "trash": 0,
    "auto-draft": 0
}
*/

```

Also add a custom post status and ensure it appears in the response:

```php
/**
 * Create initial post types.
 */
function create_initial_post_status_test() {
	register_post_status( 'ridgy_didge', array( 'public' => true ) );
}

add_action( 'init', 'create_initial_post_status_test', 0 ); // Highest priority
```

```js

await wp.apiFetch( { path: '/wp/v2/pages/count' } )

/*
{
    "publish": 1,
    "future": 0,
    "draft": 2,
    "pending": 0,
    "private": 0,
    "trash": 0,
    "ridgy_didge": 0,
    "auto-draft": 0
}
*/

```
